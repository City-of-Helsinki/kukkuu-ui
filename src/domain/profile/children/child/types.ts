import type { UpdateChildMutationInput } from '../../../api/generatedTypes/graphql';

export type ChildDetailEditModalPayload = Omit<UpdateChildMutationInput, 'id'>;
