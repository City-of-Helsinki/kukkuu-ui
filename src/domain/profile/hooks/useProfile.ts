import { useQuery, QueryResult as GenericQueryResult } from '@apollo/client';
import { useDispatch, useSelector } from 'react-redux';
import * as Sentry from '@sentry/browser';
import { useHistory } from 'react-router';
import { useTranslation } from 'react-i18next';

import { isAuthenticatedSelector } from '../../auth/state/AuthenticationSelectors';
import {
  profileQuery as ProfileQueryType,
  profileQuery_myProfile as Profile,
} from '../../api/generatedTypes/profileQuery';
import { clearEvent, saveChildrenEvents } from '../../event/state/EventActions';
import profileQuery from '../queries/ProfileQuery';
import { clearProfile, saveProfile } from '../state/ProfileActions';
import { defaultProfileData } from '../state/ProfileReducers';

export type ProfileQueryResult = Omit<
  GenericQueryResult<ProfileQueryType>,
  'data'
> & {
  data: Profile | null | undefined;
};

function useProfile(): ProfileQueryResult {
  const { i18n } = useTranslation();
  const history = useHistory();
  const dispatch = useDispatch();
  const isAuthenticated = useSelector(isAuthenticatedSelector);

  const result = useQuery<ProfileQueryType>(profileQuery, {
    skip: !isAuthenticated,
    onCompleted: (data) => {
      // Sync data to redux. Note that the redux state won't be updated
      // when apollo re-fetches queries based on refetchQueries. It's
      // better to source data from Apollo instead.
      dispatch(saveProfile(data?.myProfile || defaultProfileData));
      dispatch(clearEvent());
      dispatch(saveChildrenEvents(data?.myProfile?.children || undefined));

      // If the user has no profile it means that they have not yet
      // registered to kukkuu. In this case we want to redirect them
      // into the landing page where they can start the registration
      // process.

      // This query should be skipped when the user is not
      // authenticated. However, it seems that this does not always
      // hold and we have to check authentication again in the callback.
      if (isAuthenticated && !data?.myProfile) {
        history.replace(`/${i18n.language}/home#register`);
      }
    },
    onError: (error) => {
      // eslint-disable-next-line no-console
      console.error(error);
      dispatch(clearProfile());
      Sentry.captureException(error);
    },
  });

  return { ...result, data: result.data?.myProfile };
}

export default useProfile;
