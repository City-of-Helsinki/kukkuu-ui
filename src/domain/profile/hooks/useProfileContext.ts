import React from 'react';

import { ProfileContext } from '../ProfileContext';

export const useProfileContext = () => {
  const context = React.useContext(ProfileContext);
  if (!context) {
    throw new Error('ProfileContext must be used within a ProfileProvider');
  }
  return context;
};
