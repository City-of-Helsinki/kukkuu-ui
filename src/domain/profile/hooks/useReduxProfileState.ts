import { useDispatch, useSelector } from 'react-redux';
import { ApolloError } from '@apollo/client';
import React from 'react';
import { useOidcClient } from 'hds-react';
import * as Sentry from '@sentry/browser';

import { ProfileQuery } from '../../api/generatedTypes/graphql';
import { clearEvent, saveChildrenEvents } from '../../event/state/EventActions';
import { clearProfile, saveProfile } from '../state/ProfileActions';
import { defaultProfileData } from '../state/ProfileReducers';
import { persistor } from '../../app/state/AppStore';
import { profileSelector } from '../state/ProfileSelectors';

/**
 * Store the Guardian Profile to the redux store.
 * @deprecated the redux profile should be removed in favor of ProfileContext.
 * @returns the onSuccess and onError handlers for Apollo client query.
 */
function useReduxProfileState() {
  const dispatch = useDispatch();
  const { isAuthenticated } = useOidcClient();
  // TODO: Get rid of profile redux storage
  const reduxStorageProfile = useSelector(profileSelector);

  // TODO: Get rid of profile redux storage
  React.useEffect(() => {
    if (reduxStorageProfile?.id && !isAuthenticated()) {
      // eslint-disable-next-line no-console
      console.info('Clearing profile from redux');
      clearProfile();
      persistor
        .purge()
        .then(() => {
          // eslint-disable-next-line no-console
          console.info('Redux-persistor purged');
        })
        .catch((error) => {
          // eslint-disable-next-line no-console
          console.error(error);
        });
    }
  }, [isAuthenticated, reduxStorageProfile?.id]);

  const onSuccess = (data: ProfileQuery) => {
    // Sync data to redux. Note that the redux state won't be updated
    // when apollo re-fetches queries based on refetchQueries. It's
    // better to source data from Apollo instead.
    dispatch(saveProfile(data?.myProfile || defaultProfileData));
    dispatch(clearEvent());
    dispatch(saveChildrenEvents(data?.myProfile?.children || undefined));
  };

  const onError = (error: ApolloError) => {
    Sentry.captureException(error);
    dispatch(clearProfile());
  };

  return { onSuccess, onError };
}

export default useReduxProfileState;
