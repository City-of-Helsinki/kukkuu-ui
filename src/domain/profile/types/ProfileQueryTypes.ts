import { ProfileQuery } from '../../api/generatedTypes/graphql';

export type MyProfile = NonNullable<ProfileQuery['myProfile']>;

export type MyProfileChildren = MyProfile['children'];

export type MyProfileChild = NonNullable<
  NonNullable<MyProfileChildren['edges'][number]>['node']
>;

export type MyProfileEnrolment = NonNullable<
  NonNullable<MyProfileChild['enrolments']['edges'][number]>['node']
>;

export type LanguagesSpokenAtHomeNode = NonNullable<
  MyProfile['languagesSpokenAtHome']['edges'][number]
>['node'];

export type Project = MyProfileChild['project'];
