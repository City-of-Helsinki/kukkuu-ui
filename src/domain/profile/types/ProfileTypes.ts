export default interface ProfileValues {
  firstName: string;
}
