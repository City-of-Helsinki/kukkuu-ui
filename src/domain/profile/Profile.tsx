import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { Redirect } from 'react-router-dom';
import * as Sentry from '@sentry/browser';
import { IconPen } from 'hds-react';

import LoadingSpinner from '../../common/components/spinner/LoadingSpinner';
import ErrorMessage from '../../common/components/error/Error';
import Button from '../../common/components/button/Button';
import GiveFeedbackButton from '../../common/components/giveFeedbackButton/GiveFeedbackButton';
import Text from '../../common/components/text/Text';
import useGetPathname from '../../common/route/utils/useGetPathname';
import ListPageLayout from '../app/layout/ListPageLayout';
import { clearProfile } from './state/ProfileActions';
import useProfile from './hooks/useProfile';
import ProfileChildrenList from './children/ProfileChildrenList';
import EditProfileModal from './modal/EditProfileModal';

const Profile = () => {
  const [isOpen, setIsOpen] = useState(false);
  const { loading, error, data } = useProfile();
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const getPathname = useGetPathname();

  if (loading) return <LoadingSpinner isLoading={true} />;
  if (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    dispatch(clearProfile());
    Sentry.captureException(error);
    return <ErrorMessage message={t('api.errorMessage')} />;
  }

  if (!data) {
    // User has logged in, but not created a profile, send them to front page for registration.
    return <Redirect to={getPathname('/')} />;
  }

  return (
    <ListPageLayout>
      <ListPageLayout.Header
        title={t('profile.message.greetings', {
          firstName: data.firstName,
        })}
        content={
          <Text variant="body-l">
            {t('profile.message.serviceDescription')}
          </Text>
        }
        actions={
          <Button
            variant="secondary"
            iconLeft={<IconPen />}
            onClick={() => setIsOpen(true)}
          >
            {t('profile.edit.button.text')}
          </Button>
        }
      />
      {isOpen && (
        <EditProfileModal
          isOpen={isOpen}
          setIsOpen={setIsOpen}
          initialValues={data}
        />
      )}
      <ProfileChildrenList />
      <GiveFeedbackButton />
    </ListPageLayout>
  );
};

export default Profile;
