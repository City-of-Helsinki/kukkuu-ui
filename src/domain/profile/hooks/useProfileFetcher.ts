import React from 'react';
import { useLazyQuery } from '@apollo/client';
import * as Sentry from '@sentry/browser';

import { useIsAuthorizationReady } from '../../auth/useIsAuthorizationReady';
import useReduxProfileState from './useReduxProfileState';
import {
  ProfileQuery,
  ProfileQueryDocument,
} from '../../api/generatedTypes/graphql';
import type { UpdateProfileType } from '../ProfileContext';

function useProfileFetcher({
  setProfileToContext,
}: {
  setProfileToContext: UpdateProfileType;
}) {
  const [isFullyCalled, setIsFullyCalled] = React.useState(false);
  const [isLoginReady, loginInProgress] = useIsAuthorizationReady();

  // TODO: Get rid of redux store
  const { onSuccess: onSuccessForRedux, onError: onErrorForRedux } =
    useReduxProfileState();

  const [fetchProfile, { loading, refetch, called }] =
    useLazyQuery<ProfileQuery>(ProfileQueryDocument, {
      onCompleted: (data) => {
        setProfileToContext(data?.myProfile || null);
        onSuccessForRedux(data);
        // eslint-disable-next-line no-console
        console.info('Guardian profile fetched and set in the context.');
        setIsFullyCalled(true);
      },
      onError: (error) => {
        // eslint-disable-next-line no-console
        console.error(error);
        setProfileToContext(null);
        setIsFullyCalled(true);
        onErrorForRedux(error);
        Sentry.captureException(error);
      },
    });

  const isLoading = Boolean(loginInProgress || loading);

  // Fetch the profile information from the Kukkuu API
  // when the login provider has authenticated successfully to serve the API tokens.
  React.useEffect(() => {
    if (!called && isLoginReady) {
      // eslint-disable-next-line no-console
      console.info('Fetching profile from Kukkuu API...');
      fetchProfile();
    }
    return () => {
      setIsFullyCalled(false);
      // eslint-disable-next-line no-console
      console.info('Cleaned the fetchProfile effect.');
    };
  }, [called, fetchProfile, isLoginReady]);

  return {
    loading: isLoading,
    refetch,
    called: isFullyCalled,
  };
}

export default useProfileFetcher;
