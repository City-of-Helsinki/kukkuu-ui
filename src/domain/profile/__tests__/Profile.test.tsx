import * as HdsReact from 'hds-react';
import * as ReactRouterDom from 'react-router-dom';

import { render, waitFor } from '../../../common/test/testingLibraryUtils';
import Profile from '../Profile';
import * as ProfileContext from '../hooks/useProfileContext';
import { MyProfile } from '../types/ProfileQueryTypes';
import { Language } from '../../api/generatedTypes/graphql';

const profileMock: MyProfile = {
  id: '',
  firstName: 'Test',
  lastName: 'Guy',
  email: 'test@kukkuu.fi',
  phoneNumber: '',
  language: Language.Fi,
  hasAcceptedMarketing: false,
  children: {
    __typename: undefined,
    edges: [],
  },
  languagesSpokenAtHome: {
    __typename: undefined,
    edges: [],
  },
};

const navigateMock = vi.fn();
vi.mock('react-router-dom', async (importOriginal: any) => {
  const mod = await importOriginal();
  return {
    ...mod,
  };
});

describe('Profile', () => {
  it('renders profile correctly', () => {
    vi.spyOn(HdsReact, 'useOidcClient').mockImplementation(
      () =>
        ({
          isAuthenticated: () => true,
          isRenewing: () => false,
        }) as any
    );
    vi.spyOn(ProfileContext, 'useProfileContext').mockImplementation(
      () =>
        ({
          profile: () => profileMock,
        }) as any
    );
    const { container } = render(<Profile />);
    expect(container).toMatchSnapshot();
  });

  it('redirects to the register section in home page if user is authenticated but has no profile', async () => {
    vi.spyOn(ReactRouterDom, 'useNavigate').mockReturnValue(navigateMock);
    vi.spyOn(HdsReact, 'useOidcClient').mockImplementation(
      () =>
        ({
          isAuthenticated: () => true,
          isRenewing: () => false,
        }) as any
    );
    vi.spyOn(ProfileContext, 'useProfileContext').mockImplementation(
      () =>
        ({
          profile: null,
          isLoading: false,
          isFetchCalled: true,
        }) as any
    );

    render(<Profile />);
    await waitFor(
      () => {
        expect(navigateMock).toHaveBeenCalledWith('/home#register', {
          replace: true,
        });
      },
      { timeout: 15000 }
    );
  });
});
