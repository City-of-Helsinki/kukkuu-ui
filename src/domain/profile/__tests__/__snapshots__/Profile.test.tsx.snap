// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile > renders profile correctly 1`] = `
<div>
  <div
    class="_wrapper_70bd39"
  >
    <div
      class="Container-module_container__18lps container_hds-container__ZG2BQ _container_70bd39"
      role="main"
    >
      <div
        class="_header_70bd39"
      >
        <h1
          class="_text_e4e173 _h1_e4e173 _headerTitle_70bd39"
        >
          Hei 
        </h1>
        <div
          class="_headerContent_70bd39"
        >
          <p
            class="_text_e4e173 _body-l_e4e173 "
          >
            Hienoa että olet mukana Kulttuurin kummilapset -toiminnassa! Täällä voit tarkastella tapahtumakutsuja ja ilmoittautua tapahtumiin. Kummilapsi voi osallistua vuosittain kahteen tapahtumaan, yhteen tapahtumaan keväällä ja yhteen tapahtumaan syksyllä. Joskus tapahtumia voi olla tarjolla enemmän, silloin ilmoitamme siitä erikseen. Nähdään kulttuurin parissa!
          </p>
        </div>
        <div
          class="_headerActions_70bd39"
        >
          <button
            class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO"
            style="--color: var(--color-black); --color-hover: var(--color-black); --color-focus: var(--color-black-90); --color-hover-focus: var(--color-black-90); --background-color: transparent; --background-color-hover: var(--color-white); --background-color-focus: var(--color-white); --background-color-hover-focus: var(--color-white); --background-selected: var(--color-summer); --border-color: var(--color-summer); --border-color-hover: var(--color-summer-dark); --border-color-focus: var(--color-summer-dark); --border-color-hover-focus: var(--color-summer-dark); --border-color-selected: var(--color-summer); --border-color-selected-hover: var(--color-summer-dark); --border-color-selected-focus: var(--color-summer); --border-color-selected-hover-focus: var(--color-summer);"
            type="button"
          >
            <div
              aria-hidden="true"
              class="Button-module_icon__O-h7R button_hds-icon__17j8Z"
            >
              <svg
                aria-hidden="true"
                aria-label="pen"
                class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
                role="img"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M16.75 3L4.5 15.25L4.5143 19.9857L9.25 20L21.5 7.75L16.75 3ZM6.5 16L16.75 5.75L18.75 7.75L8.5 18L6.50696 17.994L6.5 16Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <span
              class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
            >
              Muokkaa tietojasi
            </span>
          </button>
        </div>
      </div>
      <section
        class="_layout_d7f7d5"
      >
        <h2
          class="_text_e4e173 _h2_e4e173 "
        >
          Kummilapset
        </h2>
        <div
          class="_noChild_d7f7d5"
        >
          <p>
            Ei kummilapsia
          </p>
        </div>
        <button
          aria-label="Lisää lapsi"
          class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO"
          style="--color: var(--color-black); --color-hover: var(--color-black); --color-focus: var(--color-black-90); --color-hover-focus: var(--color-black-90); --background-color: transparent; --background-color-hover: var(--color-white); --background-color-focus: var(--color-white); --background-color-hover-focus: var(--color-white); --background-selected: var(--color-summer); --border-color: var(--color-summer); --border-color-hover: var(--color-summer-dark); --border-color-focus: var(--color-summer-dark); --border-color-hover-focus: var(--color-summer-dark); --border-color-selected: var(--color-summer); --border-color-selected-hover: var(--color-summer-dark); --border-color-selected-focus: var(--color-summer); --border-color-selected-hover-focus: var(--color-summer);"
          type="button"
        >
          <div
            aria-hidden="true"
            class="Button-module_icon__O-h7R button_hds-icon__17j8Z"
          >
            <svg
              aria-hidden="true"
              aria-label="plus"
              class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M13 6V11H18V13H13V18H11V13H6V11H11V6H13Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <span
            class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
          >
            Lisää lapsi
          </span>
        </button>
      </section>
      <div
        class="_container_256fb1"
      >
        <a
          class="hds-button hds-button--secondary hds-button__label hds-button--theme-black _buttonLink_256fb1"
          href="https://docs.google.com/forms/d/e/1FAIpQLSdqw2Lq3qooEeRdgr0sV0-Wv-4XcV7IZVzq1HuWoLRa2M7tEg/viewform?usp=pp_url&entry.1982410290=Kulttuurin+kummilapset"
        >
          Anna palautetta Kulttuurin kummilapset -palvelusta
        </a>
      </div>
    </div>
  </div>
</div>
`;
